lib/net/ssh/known_hosts/cached.rb

```diff
@@ -0,0 +1,125 @@
+
+          HostKeys.new(entries.map(&:key), host, self, options)
+        end
+
+        def build_cache
+          @host_lookups = {}
+          @hmac_lookups = {}
+          @pattern_lookups = {}
+          (@user_files + @global_files).each do |file|
+            parse_known_hosts(File.expand_path(file))
+          end
+        end
+
+        def parse_known_hosts(source)
+          return unless File.readable?(source)
+
+          File.open(source) do |file|
+            @md5_sums[source] = Digest::MD5.hexdigest(File.read(file))
+            file.each_line do |line|
+              hosts, type, key_content = line.split(' ')
+              # Skip empty line or one that is commented
+              next if hosts.nil? || hosts.start_with?('#')
+
+              hostlist = hosts.split(',')
+
+              next unless KnownHosts::SUPPORTED_TYPE.include?(type)
+
+              blob = key_content.unpack("m*").first
+              key = Net::SSH::Buffer.new(blob).read_key
+
+              hostlist.each do |host|
+                entry = Entry.new(hostlist, key)
+                if host.include?('*') || host.include?('?')
+                  regex = regexify(host)
+                  @pattern_lookups[regex] ||= []
+                  @pattern_lookups[regex] << entry
+                elsif host =~ /\A\|1(\|.+){2}\z/
+                  chunks = host.split('|')
+                  salt = Base64.decode64(chunks[2])
+                  hmac = chunks[3]
+                  cache_key = [hmac, salt]
+                  @hmac_lookups[cache_key] ||= []
+                  @hmac_lookups[cache_key] << entry
+                else
+                  @host_lookups[host] ||= []
+                  @host_lookups[host] << entry
+                end
+              end
+            end
+          end
+        end
+
+        def add(host, key, _options = nil)
+          @host_lookups[host] ||= []
+          @host_lookups[host] << Entry.new([host], key)
+
+          @user_files.each do |file|
+            begin
+              # If this is the only modification since last read, update @md5_sums to preserve cache
+              preserve_cache = File.mtime(file) == @md5_sums[file]
+              KnownHosts.new(file).add(host, key)
+              @md5_sums[file] = Digest::MD5.hexdigest(File.read(file)) if preserve_cache
+              break
+            rescue SystemCallError
+            end
+          end
+        end
+
+        def regexify(pattern)
+          # see man 8 sshd for pattern details
+          pattern_regexp = pattern.split('*').map do |x|
+            x.split('?').map do |y|
+              Regexp.escape(y)
+            end.join('.')
+          end.join('[^.]*')
+
+          Regexp.new("\\A#{pattern_regexp}\\z")
+        end
+
+        def cache_invalid?
+          (@user_files + @global_files).each do |file|
+            return true if File.readable?(file) && @md5_sums[file] != Digest::MD5.hexdigest(File.read(file))
```

Comment:
I think it also needs to check if the file is removed, that should invalidate cache as well

Reply:
Updated to invalidate on delete.